pkg/build/strategies/sti/postexecutorstep.go

```diff
@@ -0,0 +1,467 @@
+	if step.builder.incremental && step.builder.config.RemovePreviousImage {
+		glog.V(1).Info("Executing step: remove previous image")
+		step.removePreviousImage(ctx.previousImageID)
+	} else {
+		glog.V(1).Info("Skipping step: remove previous image")
+	}
+	return nil
+}
+
+func (step *removePreviousImageStep) removePreviousImage(previousImageID string) {
+	if previousImageID == "" {
+		return
+	}
+
+	glog.V(1).Infof("Removing previously-tagged image %s", previousImageID)
+	if err := step.docker.RemoveImage(previousImageID); err != nil {
+		glog.V(0).Infof("error: Unable to remove previous image: %v", err)
+	}
+}
+
+type commitBuilderImageStep struct {
+	builder *STI
+	docker  dockerpkg.Docker
+}
+
+func (step *commitBuilderImageStep) execute(ctx *postExecutorStepContext) error {
+	// we're committing builder image in case of using runtime image to support incremental builds
+	if len(step.builder.config.RuntimeImage) > 0 && !step.builder.incremental {
+		glog.V(1).Info("Skipping step: commit builder image")
+		return nil
+	}
+
+	glog.V(1).Info("Executing step: commit builder image")
+
+	image := step.builder.config.BuilderImage
+
+	user, err := step.docker.GetImageUser(image)
+	if err != nil {
+		return err
+	}
+
+	cmd := step.createCommandForResultingImage(ctx.location)
+
+	ctx.labels = createLabelsForResultingImage(step.builder, step.docker, image)
+
+	ctx.imageID, err = commitContainer(step.builder, step.docker, ctx.containerID, cmd, user, step.builder.env, ctx.labels)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (step *commitBuilderImageStep) createCommandForResultingImage(location string) string {
+	cmd := step.builder.scriptsURL[api.Run]
+	if strings.HasPrefix(cmd, "image://") {
+		// scripts from inside of the image, we need to strip the image part
+		// NOTE: We use path.Join instead of filepath.Join to avoid converting the
+		// path to UNC (Windows) format as we always run this inside container.
+		cmd = strings.TrimPrefix(cmd, "image://")
+	} else {
+		// external scripts, in which case we're taking the directory to which they
+		// were extracted and append scripts dir and name
+		cmd = path.Join(location, "scripts", api.Run)
+	}
+	return cmd
+}
+
+type commitRuntimeImageStep struct {
+	builder *STI
+	docker  dockerpkg.Docker
+}
+
+func (step *commitRuntimeImageStep) execute(ctx *postExecutorStepContext) error {
+	glog.V(1).Info("Executing step: commit runtime image")
+
+	image := step.builder.config.RuntimeImage
+
+	user, err := step.docker.GetImageUser(image)
+	if err != nil {
+		return err
+	}
+
+	workDir, err := step.docker.GetImageWorkdir(image)
+	if err != nil {
+		return fmt.Errorf("Couldn't get working dir of %q image: %v", image, err)
+	}
+
+	cmd := filepath.Join(workDir, api.Run)
+
+	ctx.labels = createLabelsForResultingImage(step.builder, step.docker, image)
+
+	ctx.imageID, err = commitContainer(step.builder, step.docker, ctx.containerID, cmd, user, step.builder.env, ctx.labels)
```

Comment:
i don't think you have any control over it, do you?  I thought whatever env was passed to the container when it was launched, is going to be in any image you commit from that container.

that's why our docs currently say:

> You can add environment variables to the sourceStrategy definition of the BuildConfig. The environment variables defined there are visible during the assemble script execution and will be defined in the output image, making them also available to the run script and application code.

but yes, if you can avoid having the assemble-time env variables end up in the final image, that would be preferred.